src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/MappingService.kt

```diff
@@ -18,44 +17,40 @@ class MappingService(
   private val visitIdRepository: VisitIdRepository,
   private val telemetryClient: TelemetryClient,
 ) {
-  companion object {
+  private companion object {
     private val log = LoggerFactory.getLogger(this::class.java)
   }
 
   @Transactional
-  fun createVisitMapping(createMappingRequest: MappingDto) {
+  fun createVisitMapping(createMappingRequest: MappingDto): Mono<Void> =
     with(createMappingRequest) {
       visitIdRepository.findById(nomisId)
-        .ifPresent { throw ValidationException("Nomis visit id = $nomisId already exists") }
-      if (visitIdRepository.findOneByVsipId(vsipId) != null) {
-        throw ValidationException("VSIP visit id=$vsipId already exists")
-      }
-      visitIdRepository.save(VisitId(nomisId, vsipId, label, MappingType.valueOf(mappingType)))
-
-      telemetryClient.trackEvent(
-        "visit-created",
-        mapOf(
-          "nomisVisitId" to nomisId.toString(),
-          "vsipVisitId" to vsipId,
-          "batchId" to label,
-        ),
-        null
-      )
-      log.debug("Mapping created with VSIP visit id = $vsipId, Nomis visit id = $nomisId")
+        .map { throw ValidationException("Nomis visit id = $nomisId already exists") }
+        .thenMany(visitIdRepository.findOneByVsipId(vsipId))
+        .map { throw ValidationException("VSIP visit id=$vsipId already exists") }
+        .thenMany(visitIdRepository.save(VisitId(nomisId, vsipId, label, MappingType.valueOf(mappingType))))
+        .map {
```

Comment:
I wonder if there is some sort of `run` rather than` map` - since you want to run this after the event is published but you don't need to transform the response

Reply:
will have another look, maybe can do another `thenMany`

Reply:
found a `doOnComplete` - see what you think

Reply:
Yes `doOnComplete` reads better I think